services/httpd/handler.go

```diff
@@ -990,105 +998,82 @@ func (h *Handler) servePromRead(w http.ResponseWriter, r *http.Request, user met
-			return
-		}
-	}
-
-	opts := query.ExecutionOptions{
-		Database:  db,
-		ChunkSize: DefaultChunkSize,
-		ReadOnly:  true,
-	}
-
-	if h.Config.AuthEnabled {
-		// The current user determines the authorized actions.
-		opts.Authorizer = user
-	} else {
-		// Auth is disabled, so allow everything.
-		opts.Authorizer = query.OpenAuthorizer
-	}
-
-	// Make sure if the client disconnects we signal the query to abort
-	closing := make(chan struct{})
-	if notifier, ok := w.(http.CloseNotifier); ok {
-		// CloseNotify() is not guaranteed to send a notification when the query
-		// is closed. Use this channel to signal that the query is finished to
-		// prevent lingering goroutines that may be stuck.
-		done := make(chan struct{})
-		defer close(done)
-
-		notify := notifier.CloseNotify()
-		go func() {
-			// Wait for either the request to finish
-			// or for the client to disconnect
-			select {
-			case <-done:
-			case <-notify:
-				close(closing)
-			}
-		}()
-		opts.AbortCh = done
-	} else {
-		defer close(closing)
+	ctx := context.Background()
+	rs, err := h.Store.Read(ctx, readRequest)
+	if err != nil {
+		h.httpError(w, err.Error(), http.StatusBadRequest)
+		return
 	}
-
-	// Execute query.
-	results := h.QueryExecutor.ExecuteQuery(q, opts, closing)
+	defer rs.Close()
 
 	resp := &remote.ReadResponse{
 		Results: []*remote.QueryResult{{}},
 	}
-
-	// pull all results from the channel
-	for r := range results {
-		// Ignore nil results.
-		if r == nil {
+	for rs.Next() {
+		cur := rs.Cursor()
+		if cur == nil {
+			// no data for series key + field combination
 			continue
 		}
 
-		// read the series data and convert into Prometheus samples
-		for _, s := range r.Series {
-			ts := &remote.TimeSeries{
-				Labels: prometheus.TagsToLabelPairs(s.Tags),
-			}
+		tags := prometheus.RemoveInfluxSystemTags(rs.Tags())
+		var unsupportedCursor string
+		switch cur := cur.(type) {
+		case tsdb.FloatBatchCursor:
+			var series *remote.TimeSeries
+			for {
+				ts, vs := cur.Next()
+				if len(ts) == 0 {
+					break
+				}
 
-			for _, v := range s.Values {
-				t, ok := v[0].(time.Time)
-				if !ok {
-					h.httpError(w, fmt.Sprintf("value %v wasn't a time", v[0]), http.StatusBadRequest)
-					return
+				// We have some data for this series.
+				if series == nil {
+					series = &remote.TimeSeries{
+						Labels: prometheus.ModelTagsToLabelPairs(tags),
+					}
 				}
-				val, ok := v[1].(float64)
-				if !ok {
-					h.httpError(w, fmt.Sprintf("value %v wasn't a float64", v[1]), http.StatusBadRequest)
+
+				for i, ts := range ts {
+					series.Samples = append(series.Samples, &remote.Sample{
+						TimestampMs: ts / int64(time.Millisecond),
+						Value:       vs[i],
+					})
 				}
-				timestamp := t.UnixNano() / int64(time.Millisecond) / int64(time.Nanosecond)
-				ts.Samples = append(ts.Samples, &remote.Sample{
-					TimestampMs: timestamp,
-					Value:       val,
-				})
 			}
+			cur.Close()
```

Comment:
`cur.Close()` should be moved outside the `switch` to ensure all cursor types are closed or it will result in TSM file handle leaks.